app/javascript/packs/dashboardDropdowns.js

```diff
@@ -0,0 +1,126 @@
+import { initializeDropdown } from '@utilities/dropdownUtils';
+
+function getFormValues(form) {
+  const articleId = form.action.match(/\/(\d+)$/)[1];
+  const inputs = form.getElementsByTagName('input');
+  const formData = { id: articleId, article: {} };
+
+  for (let i = 0; i < inputs.length; i += 1) {
+    const input = inputs[i];
+    const name = input.getAttribute('name');
+    const value = input.getAttribute('value');
+
+    if (name.match(/\[(.*)\]/)) {
+      const key = name.match(/\[(.*)\]$/)[1];
+      formData.article[key] = value;
+    } else {
+      formData[name] = value;
+    }
+  }
+
+  return formData;
+}
+
+function toggleArchived(article, needsArchived) {
+  if (needsArchived === 'true') {
+    article.classList.add('story-archived', 'hidden');
+  } else {
+    article.classList.remove('story-archived');
+  }
+}
+
+function toggleNotifications(submit, action) {
+  if (action === 'Mute Notifications') {
+    submit.setAttribute('value', 'Receive Notifications');
+  } else {
+    submit.setAttribute('value', 'Mute Notifications');
+  }
+}
+
+function onXhrSuccess(form, article, values) {
+  if (values.article.archived) {
+    toggleArchived(article, values.article.archived);
+  } else {
+    const submit = form.querySelector('[type="submit"]');
+    const submitValue = submit.getAttribute('value');
+
+    toggleNotifications(submit, submitValue);
+  }
+}
+
+const handleFormSubmit = (e) => {
+  e.preventDefault();
+  e.stopPropagation();
+
+  const { target: form } = e;
+  const values = getFormValues(form);
+  const data = JSON.stringify(values);
+
+  const formData = new FormData(form);
+  const method = formData.get('_method') || 'post';
+
+  const xhr = new XMLHttpRequest();
+  xhr.open(method.toUpperCase(), form.action);
```

Comment:
Not that it matters, more out of curiosity: does the request method really need to be upcased? I'd find that a bit strange for a string argument that could be normalized in the method itself.

Reply:
tbh this, and the other areas you highlighted were a straight copy from the original code in initializers 😄  These functions had to be moved to the new file but aren't really related to the feature at hand. Totally makes sense to optimise them now though, so I'll take a look and incorporate the feedback - thanks!

Reply:
it turns out it does need to be uppercased! TIL